New .broadcast_like() incorrect

Unfortunately, more robust tests show there is some logic problem in the @dcherian solution to #2885.  I was merging my work in #3127 with his more-principled refactoring in #3086, and my tests showed an error in the refactoring.

In https://github.com/DavidMertz/xarray/tree/feature/2885-broadcast_like-merge-branch I have created a temporary method `.broadcast_like_naive()` and have created a test against the naive implementation that is identical to the test against the refactored code.

We actually get the failure in the call to `.broadcast_like()` itself, i.e. not in the assertion about the result:

```python
___________________________________________ TestDataArray.test_broadcast_like ____________________________________________

self = <xarray.tests.test_dataarray.TestDataArray object at 0x7f2287e3a828>

    def test_broadcast_like(self):
        arr1 = DataArray(np.ones((2, 3)), dims=['x', 'y'],
                         coords={'x': ['a', 'b'], 'y': ['a', 'b', 'c']})
        arr2 = DataArray(np.ones((3, 2)), dims=['x', 'y'],
                         coords={'x': ['a', 'b', 'c'], 'y': ['a', 'b']})
        orig1, orig2 = broadcast(arr1, arr2)
>       new1 = arr2.broadcast_like(arr1)

xarray/tests/test_dataarray.py:1276: 
[...]
E                   ValueError: conflicting sizes for dimension 'y': length 2 on the data but length 3 on coordinate 'y'
```

--- a/doc/whats-new.rst
+++ b/doc/whats-new.rst
@@ -25,7 +25,8 @@ New functions/methods
 ~~~~~~~~~~~~~~~~~~~~~
 
 - Added :py:meth:`DataArray.broadcast_like` and :py:meth:`Dataset.broadcast_like`.
-  By `Deepak Cherian <https://github.com/dcherian>`_.
+  By `Deepak Cherian <https://github.com/dcherian>`_ and `David Mertz 
+  <http://github.com/DavidMertz>`_.
 
 Enhancements
 ~~~~~~~~~~~~

--- a/xarray/core/dataarray.py
+++ b/xarray/core/dataarray.py
@@ -1063,7 +1063,7 @@ class DataArray(AbstractArray, DataWithCoords):
         dims_map, common_coords = _get_broadcast_dims_map_common_coords(
             args, exclude)
 
-        return _broadcast_helper(self, exclude, dims_map, common_coords)
+        return _broadcast_helper(args[1], exclude, dims_map, common_coords)
 
     def reindex_like(self, other: Union['DataArray', Dataset],
                      method: Optional[str] = None, tolerance=None,

--- a/xarray/core/dataarray.py
+++ b/xarray/core/dataarray.py
@@ -995,11 +995,6 @@ class DataArray(AbstractArray, DataWithCoords):
             dim=dim, method=method, tolerance=tolerance, **indexers)
         return self._from_temp_dataset(ds)
 
-    def broadcast_like_naive(self,
-                             other: Union['DataArray', Dataset],
-                             exclude=None) -> 'DataArray':
-        return broadcast(self, other)[1]
-
     def broadcast_like(self,
                        other: Union['DataArray', Dataset],
                        exclude=None) -> 'DataArray':
@@ -1033,28 +1028,27 @@ class DataArray(AbstractArray, DataWithCoords):
 
         >>> arr1
         <xarray.DataArray (x: 2, y: 3)>
-        array([[0.824093, 0.769792, 0.571621],
-               [0.310378, 0.480418, 0.062015]])
+        array([[0.840235, 0.215216, 0.77917 ],
+               [0.726351, 0.543824, 0.875115]])
         Coordinates:
           * x        (x) <U1 'a' 'b'
           * y        (y) <U1 'a' 'b' 'c'
         >>> arr2
         <xarray.DataArray (x: 3, y: 2)>
-        array([[0.852992, 0.106589],
-               [0.087549, 0.563304],
-               [0.675744, 0.285752]])
+        array([[0.612611, 0.125753],
+               [0.853181, 0.948818],
+               [0.180885, 0.33363 ]])
         Coordinates:
           * x        (x) <U1 'a' 'b' 'c'
           * y        (y) <U1 'a' 'b'
         >>> arr1.broadcast_like(arr2)
         <xarray.DataArray (x: 3, y: 3)>
-        array([[0.852992, 0.106589,      nan],
-               [0.087549, 0.563304,      nan],
-               [0.675744, 0.285752,      nan]])
+        array([[0.840235, 0.215216, 0.77917 ],
+               [0.726351, 0.543824, 0.875115],
+               [     nan,      nan,      nan]])
         Coordinates:
           * x        (x) object 'a' 'b' 'c'
           * y        (y) object 'a' 'b' 'c'
-
         """
         if exclude is None:
             exclude = set()

--- a/xarray/core/dataarray.py
+++ b/xarray/core/dataarray.py
@@ -15,7 +15,7 @@ from . import (
     utils)
 from .accessor_dt import DatetimeAccessor
 from .accessor_str import StringAccessor
-from .alignment import (align, broadcast, _broadcast_helper,
+from .alignment import (align, _broadcast_helper,
                         _get_broadcast_dims_map_common_coords,
                         reindex_like_indexers)
 from .common import AbstractArray, DataWithCoords

--- a/xarray/core/dataarray.py
+++ b/xarray/core/dataarray.py
@@ -996,8 +996,8 @@ class DataArray(AbstractArray, DataWithCoords):
         return self._from_temp_dataset(ds)
 
     def broadcast_like_naive(self,
-                       other: Union['DataArray', Dataset],
-                       exclude=None) -> 'DataArray':
+                             other: Union['DataArray', Dataset],
+                             exclude=None) -> 'DataArray':
         return broadcast(self, other)[1]
 
     def broadcast_like(self,

--- a/xarray/core/dataarray.py
+++ b/xarray/core/dataarray.py
@@ -15,7 +15,7 @@ from . import (
     utils)
 from .accessor_dt import DatetimeAccessor
 from .accessor_str import StringAccessor
-from .alignment import (align, _broadcast_helper,
+from .alignment import (align, broadcast, _broadcast_helper,
                         _get_broadcast_dims_map_common_coords,
                         reindex_like_indexers)
 from .common import AbstractArray, DataWithCoords
@@ -995,20 +995,67 @@ class DataArray(AbstractArray, DataWithCoords):
             dim=dim, method=method, tolerance=tolerance, **indexers)
         return self._from_temp_dataset(ds)
 
+    def broadcast_like_naive(self,
+                       other: Union['DataArray', Dataset],
+                       exclude=None) -> 'DataArray':
+        return broadcast(self, other)[1]
+
     def broadcast_like(self,
                        other: Union['DataArray', Dataset],
                        exclude=None) -> 'DataArray':
-        """Broadcast this DataArray against another Dataset or DataArray.
+        """Broadcast a DataArray to the shape of another DataArray or Dataset
+
         This is equivalent to xr.broadcast(other, self)[1]
+
+        xarray objects are broadcast against each other in arithmetic
+        operations, so this method is not be necessary for most uses.
+
+        If no change is needed, the input data is returned to the output
+        without being copied.
+
+        If new coords are added by the broadcast, their values are
+        NaN filled.
 
         Parameters
         ----------
         other : Dataset or DataArray
             Object against which to broadcast this array.
+
         exclude : sequence of str, optional
             Dimensions that must not be broadcasted
-        """
 
+        Returns
+        -------
+        new_da: xr.DataArray
+
+        Examples
+        --------
+
+        >>> arr1
+        <xarray.DataArray (x: 2, y: 3)>
+        array([[0.824093, 0.769792, 0.571621],
+               [0.310378, 0.480418, 0.062015]])
+        Coordinates:
+          * x        (x) <U1 'a' 'b'
+          * y        (y) <U1 'a' 'b' 'c'
+        >>> arr2
+        <xarray.DataArray (x: 3, y: 2)>
+        array([[0.852992, 0.106589],
+               [0.087549, 0.563304],
+               [0.675744, 0.285752]])
+        Coordinates:
+          * x        (x) <U1 'a' 'b' 'c'
+          * y        (y) <U1 'a' 'b'
+        >>> arr1.broadcast_like(arr2)
+        <xarray.DataArray (x: 3, y: 3)>
+        array([[0.852992, 0.106589,      nan],
+               [0.087549, 0.563304,      nan],
+               [0.675744, 0.285752,      nan]])
+        Coordinates:
+          * x        (x) object 'a' 'b' 'c'
+          * y        (y) object 'a' 'b' 'c'
+
+        """
         if exclude is None:
             exclude = set()
         args = align(other, self, join='outer', copy=False, exclude=exclude)

--- a/xarray/core/dataset.py
+++ b/xarray/core/dataset.py
@@ -2039,7 +2039,7 @@ class Dataset(Mapping, ImplementsDatasetReduce, DataWithCoords):
         dims_map, common_coords = _get_broadcast_dims_map_common_coords(
             args, exclude)
 
-        return _broadcast_helper(self, exclude, dims_map, common_coords)
+        return _broadcast_helper(args[1], exclude, dims_map, common_coords)
 
     def reindex_like(self, other, method=None, tolerance=None, copy=True,
                      fill_value=dtypes.NA):

--- a/xarray/tests/test_dataarray.py
+++ b/xarray/tests/test_dataarray.py
@@ -1255,20 +1255,29 @@ class TestDataArray:
         expected = DataArray(2, coords={1: 2}, name=1)
         assert_identical(actual, expected)
 
+    def test_broadcast_like_naive(self):
+        arr1 = DataArray(np.ones((2, 3)), dims=['x', 'y'],
+                         coords={'x': ['a', 'b'], 'y': ['a', 'b', 'c']})
+        arr2 = DataArray(np.ones((3, 2)), dims=['x', 'y'],
+                         coords={'x': ['a', 'b', 'c'], 'y': ['a', 'b']})
+        orig1, orig2 = broadcast(arr1, arr2)
+        new1 = arr2.broadcast_like_naive(arr1)
+        new2 = arr1.broadcast_like_naive(arr2)
+      
+        assert orig1.identical(new1)
+        assert orig2.identical(new2)
+
     def test_broadcast_like(self):
-        original1 = DataArray(np.random.randn(5),
-                              [('x', range(5))])
-
-        original2 = DataArray(np.random.randn(6),
-                              [('y', range(6))])
-
-        expected1, expected2 = broadcast(original1, original2)
-
-        assert_identical(original1.broadcast_like(original2),
-                         expected1.transpose('y', 'x'))
-
-        assert_identical(original2.broadcast_like(original1),
-                         expected2)
+        arr1 = DataArray(np.ones((2, 3)), dims=['x', 'y'],
+                         coords={'x': ['a', 'b'], 'y': ['a', 'b', 'c']})
+        arr2 = DataArray(np.ones((3, 2)), dims=['x', 'y'],
+                         coords={'x': ['a', 'b', 'c'], 'y': ['a', 'b']})
+        orig1, orig2 = broadcast(arr1, arr2)
+        new1 = arr2.broadcast_like(arr1)
+        new2 = arr1.broadcast_like(arr2)
+      
+        assert orig1.identical(new1)
+        assert orig2.identical(new2)
 
     def test_reindex_like(self):
         foo = DataArray(np.random.randn(5, 6),

--- a/xarray/tests/test_dataarray.py
+++ b/xarray/tests/test_dataarray.py
@@ -1267,10 +1267,8 @@ class TestDataArray:
         assert orig1.identical(new1)
         assert orig2.identical(new2)
 
-        orig3 = DataArray(np.random.randn(5),
-                              [('x', range(5))])
-        orig4 = DataArray(np.random.randn(6),
-                              [('y', range(6))])
+        orig3 = DataArray(np.random.randn(5), [('x', range(5))])
+        orig4 = DataArray(np.random.randn(6), [('y', range(6))])
         new3, new4 = broadcast(orig3, orig4)
 
         assert_identical(orig3.broadcast_like(orig4), new3.transpose('y', 'x'))

--- a/xarray/tests/test_dataarray.py
+++ b/xarray/tests/test_dataarray.py
@@ -1267,6 +1267,15 @@ class TestDataArray:
         assert orig1.identical(new1)
         assert orig2.identical(new2)
 
+        orig3 = DataArray(np.random.randn(5),
+                              [('x', range(5))])
+        orig4 = DataArray(np.random.randn(6),
+                              [('y', range(6))])
+        new3, new4 = broadcast(orig3, orig4)
+
+        assert_identical(orig3.broadcast_like(orig4), new3.transpose('y', 'x'))
+        assert_identical(orig4.broadcast_like(orig3), new4)
+
     def test_reindex_like(self):
         foo = DataArray(np.random.randn(5, 6),
                         [('x', range(5)), ('y', range(6))])

--- a/xarray/tests/test_dataarray.py
+++ b/xarray/tests/test_dataarray.py
@@ -1263,7 +1263,7 @@ class TestDataArray:
         orig1, orig2 = broadcast(arr1, arr2)
         new1 = arr2.broadcast_like_naive(arr1)
         new2 = arr1.broadcast_like_naive(arr2)
-      
+ 
         assert orig1.identical(new1)
         assert orig2.identical(new2)
 
@@ -1275,7 +1275,7 @@ class TestDataArray:
         orig1, orig2 = broadcast(arr1, arr2)
         new1 = arr2.broadcast_like(arr1)
         new2 = arr1.broadcast_like(arr2)
-      
+
         assert orig1.identical(new1)
         assert orig2.identical(new2)
 

--- a/xarray/tests/test_dataarray.py
+++ b/xarray/tests/test_dataarray.py
@@ -1255,26 +1255,14 @@ class TestDataArray:
         expected = DataArray(2, coords={1: 2}, name=1)
         assert_identical(actual, expected)
 
-    def test_broadcast_like_naive(self):
-        arr1 = DataArray(np.ones((2, 3)), dims=['x', 'y'],
-                         coords={'x': ['a', 'b'], 'y': ['a', 'b', 'c']})
-        arr2 = DataArray(np.ones((3, 2)), dims=['x', 'y'],
-                         coords={'x': ['a', 'b', 'c'], 'y': ['a', 'b']})
-        orig1, orig2 = broadcast(arr1, arr2)
-        new1 = arr2.broadcast_like_naive(arr1)
-        new2 = arr1.broadcast_like_naive(arr2)
-
-        assert orig1.identical(new1)
-        assert orig2.identical(new2)
-
     def test_broadcast_like(self):
         arr1 = DataArray(np.ones((2, 3)), dims=['x', 'y'],
                          coords={'x': ['a', 'b'], 'y': ['a', 'b', 'c']})
         arr2 = DataArray(np.ones((3, 2)), dims=['x', 'y'],
                          coords={'x': ['a', 'b', 'c'], 'y': ['a', 'b']})
         orig1, orig2 = broadcast(arr1, arr2)
-        new1 = arr2.broadcast_like(arr1)
-        new2 = arr1.broadcast_like(arr2)
+        new1 = arr1.broadcast_like(arr2)
+        new2 = arr2.broadcast_like(arr1)
 
         assert orig1.identical(new1)
         assert orig2.identical(new2)

--- a/xarray/tests/test_dataarray.py
+++ b/xarray/tests/test_dataarray.py
@@ -1263,7 +1263,7 @@ class TestDataArray:
         orig1, orig2 = broadcast(arr1, arr2)
         new1 = arr2.broadcast_like_naive(arr1)
         new2 = arr1.broadcast_like_naive(arr2)
- 
+
         assert orig1.identical(new1)
         assert orig2.identical(new2)
 

